src/socket/address.cr

```diff
@@ -144,16 +144,111 @@ class Socket
       parse URI.parse(uri)
     end
 
+    # Returns the IPv4 address with the given address *fields* and *port*
+    # number.
+    def self.v4(fields : UInt8[4], port : UInt16) : self
+      addr_value = UInt32.zero
+      fields.each_with_index do |field, i|
+        addr_value = (addr_value << 8) | field
+      end
+
+      addr = LibC::SockaddrIn.new(
+        sin_family: LibC::AF_INET,
+        sin_port: endian_swap(port),
+        sin_addr: LibC::InAddr.new(s_addr: endian_swap(addr_value)),
+      )
+      new(pointerof(addr), sizeof(typeof(addr)))
```

Comment:
@HertzDevil I just noticed this now during the review of #13463.
Taking a pointer to a variable on the stack is dangerous. It seems to be working here, but it's fragile.
It probably only works because `.new` and `#initialize` don't allocate enough variables to override the (relevant) parts on the stack.
I suppose a safer alternative could perhaps be a yielding variant of `#initialize` or some other way to directly assign to the memory of the `Address` struct.
Same applies for `v6`.

Reply:
There isn't a tail call here (I believe you have to opt in explicitly in the LLVM IR), so `addr` will be valid before `new` returns.

In any case I plan to make `SockaddrIn` or `SockaddrIn6` part of the `IPAddress` itself so that `#to_unsafe` never allocates. That would allow `addr` to be passed by value here

Reply:
Okay, if LLVM keeps this safe, then it's fine.